fairscale/nn/misc/offload.py

```diff
@@ -125,44 +125,74 @@ class ShardSyncLayer(torch.autograd.Function):
-        # Drop the shard we just went through, except if this is the last one in line
-        if layer_window[1] and layer_window[2]:
-            layer_window[1].forward_drop(non_blocking=True)
-
-        # Start the load of the next shard in line, opportunistically look ahead
-        if layer_window[3]:
-            layer_window[3].forward_load(non_blocking=True)
-
-        ctx.layer_window = layer_window
+    def forward(ctx: Any, inputs: Any, index: int, model_slices: Any, model_instance: Any) -> Any:  # type: ignore
+        drop_index = index
+        load_index = index + 1
+        max_slices = len(model_slices)
+
+        if drop_index >= 0:
+            # Move shard from device to offload device.
+            logging.info(f"Dropping shard {drop_index}")
+            model_slices[drop_index].forward_drop()
+
+        if load_index < max_slices:
+            # Load shard from offload device to device.
+            logging.info(f"Loading shard{load_index}")
+            model_slices[load_index].forward_load()
+
+        ctx.inputs = inputs
+        ctx.index = index
+        ctx.model_slices = model_slices
+        ctx.model_instance = model_instance
 
         return inputs if isinstance(inputs, tuple) else (inputs,)
 
     @staticmethod
     def backward(ctx, *grad_outputs):  # type: ignore
-        if ctx.layer_window[3]:
-            ctx.layer_window[3].backward_drop(non_blocking=True)
 
-        # Opportunistically pre-load ahead of the compute wavefront
-        if ctx.layer_window[1]:
-            ctx.layer_window[1].backward_load(non_blocking=True)
+        load_index = ctx.index
+        drop_index = load_index + 1
+        model_slices = ctx.model_slices
+        model_instance = ctx.model_instance
+
+        # TODO(anj-s): Are these redundant in the backward pass?
+        if drop_index == len(model_slices):
+            # Drop the last activation since it is still on the CPU
+            # after the loss.backward() call.
+            model_instance._activations[-1] = tuple([a.cuda() for a in list(model_instance._activations[-1])])
+
+        if drop_index < len(model_slices):
+            # Move shard from device to offload device.
+            logging.info(f"Backward Dropping shard {drop_index}")
+            model_slices[drop_index].backward_drop()
+            model_instance._activations[drop_index] = tuple(
+                [a.cpu() for a in list(model_instance._activations[drop_index])]
+            )
+
+        if load_index >= 0:
+            # Load shard from offload device to device.
+            logging.info(f"Backward Loading shard{load_index}")
+            model_slices[load_index].backward_load()
+            model_instance._activations[load_index] = tuple(
+                [a.cuda() for a in list(model_instance._activations[load_index])]
+            )
 
         # The returned variables need to mirror the forward inputs
+        # TODO(anj-s): Why do we need to do this?
```

Comment:
not sure that it's needed anymore, at some point I was testing with another model which did not handle tuples, feel free to discard

Reply:
I think we still need it and it needs to be equal to the number of inputs. I am going to follow up on this before removing the TODO.